apple/internal/xctrunner.bzl

```
@@ -0,0 +1,157 @@
+"""
+Rule for merging multiple test targets into a single XCTRunner.app bundle.
+"""
+
+load(
+    "//apple:providers.bzl",
+    "AppleBundleInfo",
+)
+load(
+    "//apple/internal:providers.bzl",
+    "new_applebinaryinfo",
+)
+
+_TestBundleInfo = provider(
+    "Test bundle info for tests that will be run.",
+    fields = {
+        "infoplists": "A `depset` of `File`s of `Info.plist` files.",
+        "xctests": "A `depset` of paths of XCTest bundles.",
+    },
+)
+
+def _test_bundle_info_aspect_impl(target, ctx):
+    rule_attr = ctx.rule.attr
+
+    if AppleBundleInfo in target:
+        info = target[AppleBundleInfo]
+        xctests = depset([info.archive])
+        infoplists = depset([info.infoplist])
+    else:
+        deps = getattr(rule_attr, "tests", [])
+        xctests = depset(
+            transitive = [
+                dep[_TestBundleInfo].xctests
+                for dep in deps
+            ],
+        )
+        infoplists = depset(
+            transitive = [
+                dep[_TestBundleInfo].infoplists
+                for dep in deps
+            ],
+        )
+
+    return [
+        _TestBundleInfo(
+            infoplists = infoplists,
+            xctests = xctests,
+        ),
+    ]
+
+test_bundle_info_aspect = aspect(
+    attr_aspects = ["tests"],
+    implementation = _test_bundle_info_aspect_impl,
+)
+
+def _xctrunner_impl(ctx):
+    # Get test target info
+    infos = [target[_TestBundleInfo] for target in ctx.attr.test_targets]
+    infoplists = depset(
+        transitive = [info.infoplists for info in infos],
+    )
+    xctests = depset(
+        transitive = [info.xctests for info in infos],
+    )
+    output = ctx.actions.declare_directory(ctx.label.name + ".app")
+
+    # Args for _make_xctrunner
+    arguments = ctx.actions.args()
+    arguments.add("--name", ctx.label.name)
+    arguments.add("--platform", ctx.attr.platform)
+
+    if ctx.attr.verbose:
+        arguments.add("--verbose", ctx.attr.verbose)
```

Could this be made into a toggle?
```suggestion
        args.add("--verbose")
```